src/cmd/tools/verify_data_files/main/main.go

```diff
@@ -0,0 +1,473 @@
+	log *zap.Logger,
+) error {
+	fsOpts := fs.NewOptions().SetFilePathPrefix(opts.filePathPrefix)
+	reader, err := fs.NewReader(opts.bytesPool, fsOpts)
+	if err != nil {
+		return err
+	}
+
+	fileSet := opts.fileSet
+
+	openOpts := fs.DataReaderOpenOptions{
+		Identifier:  fileSet.ID,
+		FileSetType: persist.FileSetFlushType,
+	}
+
+	err = reader.Open(openOpts)
+	if err != nil {
+		return err
+	}
+
+	defer reader.Close()
+
+	for {
+		id, tags, data, checksum, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		check, err := readEntry(id, tags, data, checksum)
+		data.Finalize() // Always finalize data.
+		if err != nil {
+			shouldFixInvalidID := check.invalidID && opts.fixInvalidIDs
+			shouldFixInvalidTags := check.invalidTags && opts.fixInvalidTags
+			if shouldFixInvalidID || shouldFixInvalidTags {
+				log.Info("starting to fix file set", zap.Any("fileSet", fileSet))
+				fixErr := fixFileSet(opts, log)
+				if fixErr == nil {
+					log.Info("fixed file set", zap.Any("fileSet", fileSet))
+					return err
+				}
+
+				log.Error("could not fix file set",
+					zap.Any("fileSet", fileSet), zap.Error(fixErr))
+			}
+			return err
+		}
+	}
+
+	return nil
+}
+
+type readEntryResult struct {
+	invalidID       bool
+	invalidTags     bool
+	invalidChecksum bool
+}
+
+func readEntry(
+	id ident.ID,
+	tags ident.TagIterator,
+	data checked.Bytes,
+	checksum uint32,
+) (readEntryResult, error) {
+	if !utf8.Valid(id.Bytes()) {
+		return readEntryResult{invalidID: true},
+			fmt.Errorf("invalid id: err=%s, as_string=%s, as_hex=%x",
+				"non-utf8", id.Bytes(), id.Bytes())
+	}
+
+	for tags.Next() {
+		tag := tags.Current()
+		if err := convert.ValidateSeriesTag(tag); err != nil {
+			return readEntryResult{invalidTags: true},
+				fmt.Errorf("invalid tag: err=%v, "+
+					"name_as_string=%s, name_as_hex=%s"+
+					"value_as_string=%s, value_as_hex=%s",
+					err,
+					tag.Name.Bytes(), tag.Name.Bytes(),
+					tag.Value.Bytes(), tag.Value.Bytes())
+		}
+	}
+
+	data.IncRef()
+	calculatedChecksum := digest.Checksum(data.Bytes())
+	data.DecRef()
+
+	if calculatedChecksum != checksum {
+		return readEntryResult{invalidChecksum: true},
+			fmt.Errorf("data checksum invalid: actual=%v, expected=%v",
+				calculatedChecksum, checksum)
+	}
+	return readEntryResult{}, nil
+}
+
+func fixFileSet(
+	opts verifyFileSetOptions,
+	log *zap.Logger,
+) error {
+	fsOpts := fs.NewOptions().SetFilePathPrefix(opts.filePathPrefix)
+	reader, err := fs.NewReader(opts.bytesPool, fsOpts)
+	if err != nil {
+		return err
+	}
+
+	fileSet := opts.fileSet
+
+	openOpts := fs.DataReaderOpenOptions{
+		Identifier:  fileSet.ID,
+		FileSetType: persist.FileSetFlushType,
+	}
+
+	err = reader.Open(openOpts)
+	if err != nil {
+		return err
+	}
+
+	defer reader.Close()
+
+	// NOTE: we output to a new directory so that we don't clobber files.
+	writeFsOpts := fsOpts.SetFilePathPrefix(opts.fixDir)
+	writer, err := fs.NewWriter(writeFsOpts)
```

Comment:
Should this `defer writer.Close()` and return nil at the end instead?

Reply:
Sure thing.

Reply:
Since `writer.Close()` does quite a few final steps, I'm going to continue to return an error if it returns an error.

To make sure we always close up I changed it to this:
```go
	success := false
	defer func() {
		if !success {
			writer.Close()
		}
	}()

	// do work

	success = true
	return writer.Close()
```

Reply:
Sounds good 👍